File: [0094][Medium][Binary_Tree_Inorder_Traversal][BST]/Binary_Tree_Inorder_Traversal_2.py
# -*- coding: utf-8 -*-  
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
# solution 2, iteration with stack
class Solution(object):
	def inorderTraversal(self, root):
		"""
		:type root: TreeNode
		:rtype: List[int]
		"""
		res = []
		stack = []
		current = root
		while stack or current:
			# 不断往左子树方向走，每走一次就将当前节点保存到栈中
			# 这是模拟递归的调用
			if current:
				stack.append(current)
				current = current.left
			# 当前节点为空，说明左边走到头了，从栈中弹出节点并保存
			# 然后转向右边节点，继续上面整个过程
			else:
				tmp = stack.pop()
				res.append(tmp.val)
				current = tmp.right
		return res

File: [0094][Medium][Binary_Tree_Inorder_Traversal][BST]/Binary_Tree_Inorder_Traversal_2.java
//Definition for a binary tree node.
public class TreeNode {
    int val;
   TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

// Time  Complexity: O(N)
// Space Complexity: O(N)
// Solution 2, with stack
public class Solution {
    public List < Integer > inorderTraversal(TreeNode root) {
        List < Integer > res = new ArrayList < > ();
        Stack < TreeNode > stack = new Stack < > ();
        TreeNode curr = root;
        while (curr != null || !stack.isEmpty()) {
            while (curr != null) {
                stack.push(curr);
                curr = curr.left;
            }
            curr = stack.pop();
            res.add(curr.val);
            curr = curr.right;
        }
        return res;
    }
}